doc/ingest-proposal.md

```diff
@@ -0,0 +1,102 @@
+# Proposal for a new Ingest design
+
+## Current Problems
+
+1. Not resilient to process death.
+2. Half the logic is in storage writer the other half in a block hook.
+3. Undefined behavior if multiple publishers sync the same chain.
+4. Relies of a chain of blocking channels to reverse the order of graphsync.
+5. not linear execution flow.
+
+The indexer semantics are relatively simple, yet the ingester may fail in
+surprising and hard to reason about ways. The goal of this proposal is to
+outline a design for an alternate ingester that upholds the indexer semantics
+while being easy to reason about.
+
+## Overview of the network
+
+The indexer ingests data from publishers. There can be one publisher per
+provider or many publishers per provider. Additionally there can be faulty or
+malicious publishers that can point to other chains.
+
+* This means we shouldn't assume a 1:1 relationship with a publisher and a
+provider.
+
+![network overview](./assets/ingesterNetworkOverview.png)
+
+
+## Constraints of the indexer
+
+1. If a an Ad is applied all older ads this ad references have also been
+   applied. For example, given some chain `A <- B <- C`, the indexer will never
+   be in the state that it indexed `A & C` but not `B`.
+2. An indexer will index an Ad chain, but will not make any gaurantees about
+   consistency in the presence of multiple ad chains for a given provider. For
+   example if a provider publishes two ad chains at the same time chain1 and
+   chain2 the indexer will apply whichever chain it learns about first first,
+   then apply the other chain.
+3. An indexer will not index the same Ad twice.
+4. An indexer will be resilient to restarts. If the indexer goes down and comes
+   back up it should not break constraint 1.
+
+## Proposal
+
+1. Introduce a new concept called "staging". An advertisement chain is put into
+   the staging area after it is synced but before the chain is indexed.
+2. Each *provider* has its own staging area.
```

Comment:
What happens if there are multiple publishers concurrently syncing the same chain for the same provider?

Reply:
See the bottom half of the ingesterStaging.jpg image.
and:
> When the handler's sync is finished it moves the stack(s)[2] into each provider's staging area.

so it may put the ads that need to be processed in the staging area. And those ads may already be in the staging area, but that's okay it will just be a no-op when we process the ad.